feat: Update Event Retention doc (#1415)
edgexfoundry/edgex-docs: docs_src/microservices/core/data/details/DataRetention.md

# Core Data - Data Retention and Persistent Caps

!!! edgey "EdgeX 4.0"
    Enhance the original design in EdgeX 4.0, the Core Data service purge events by specified auto event source and retention policy.

## Overview

> In use cases, since core data persists data in the local database indefinitely, there is a need to persist the latest recent readings only and clean up the old ones, as keeping an infinite number of readings is considered computationally expensive and can lead to a lack of machine memory. Thus, a retention mechanism is placed on core data to keep events and readings.

Author: Updated.

## Configure the Retention Policy

### Define the Retention policy in the AutoEvent

> You can define the `retention policy` in the `auto event` when adding or updating the device data as below. The Core Data service will apply the default retention values if you didn't define it, and do time-based or count-based event retention according to the retention policy.

Author: Updated.

                "minCap": -1, 
                "duration": "24h"
            }
        }
    ]
}
```
In this case, the `minCap` is -1 which means we don't keep old data, then the Core Data purge events that the age(current time - event origin) is greater than "24h".

### Time-based Retention With MinCap
You can define the `duration` and `minCap` for time-based retention and also keep minimum expired events in DB. Note that the `minCap` is used to keep old data when purging events, even if the database contains new data, the process still keep old data. This is useful when the device unexpectedly shutdown, and we still want to keep some old data for tracing.
```
"device": {
    "name": "device_int_autoevents",
    ...
    "profileName": "profile_int_resources",
    "autoEvents": [
        { 
            "interval": "1s", "onChange": false, "sourceName": "INT16_1", 
            "retention": {
                "maxCap": -1, 
                "minCap": 1000, 
                "duration": "24h"
            }
        }
    ]
}
```
In this case, the Core Data service purges events that the age(current time - event origin) is greater than "24h" and leaves 1000 old events in DB. 

> For More details, the Core Data service will find the latest 1000th event, if the 1000th event doesn't exist, the service skip purging, otherwise,  the service removes the expired data that is older than the 1000th event.

Reviewer: ```suggestion
The Core Data service identifies the most recent 1000th event. If the 1000th event does not exist, the service skips purging. Otherwise, it removes expired data older than the 1000th event.
```

Author: Updated.
